File: packages/oidc_resource_server/lib/src/claims.dart
/// One FGAC tuple embedded in the access token (`fgac_relations` claim).
class FgacRelationClaim {
  const FgacRelationClaim({
    required this.resourceType,
    required this.resourceId,
    required this.relation,
  });

  final String resourceType;
  final String resourceId;
  final String relation;
}

Set<String> parseScopeClaim(Object? scope) {
  if (scope is! String || scope.isEmpty) {
    return {};
  }
  return scope.split(RegExp(r'\s+')).where((s) => s.isNotEmpty).toSet();
}

List<String> parseRealmRoles(Map<String, dynamic> claims) {
  final raw = claims['realm_access'];
  if (raw is! Map<String, dynamic>) {
    return const [];
  }
  final roles = raw['roles'];
  if (roles is! List) {
    return const [];
  }
  return roles.whereType<String>().toList();
}

Map<String, List<String>> parseResourceAccess(Map<String, dynamic> claims) {
  final raw = claims['resource_access'];
  if (raw is! Map<String, dynamic>) {
    return {};
  }
  final out = <String, List<String>>{};
  for (final e in raw.entries) {
    final v = e.value;
    if (v is! Map<String, dynamic>) continue;
    final roles = v['roles'];
    if (roles is! List) continue;
    out[e.key] = roles.whereType<String>().toList();
  }
  return out;
}

List<FgacRelationClaim> parseFgacRelations(Map<String, dynamic> claims) {
  final raw = claims['fgac_relations'];
  if (raw is! List) {
    return const [];
  }
  final out = <FgacRelationClaim>[];
  for (final item in raw) {
    if (item is! Map<String, dynamic>) continue;
    final t = item['resource_type'];
    final id = item['resource_id'];
    final r = item['relation'];
    if (t is String && id is String && r is String) {
      out.add(FgacRelationClaim(resourceType: t, resourceId: id, relation: r));
    }
  }
  return out;
}

bool parseFgacTruncated(Map<String, dynamic> claims) {
  return claims['fgac_truncated'] == true;
}

bool matchesFgacGrant(
  List<FgacRelationClaim> relations,
  String resourceType,
  String resourceId, {
  String? relation,
}) {
  for (final r in relations) {
    if (r.resourceType != resourceType) continue;
    if (relation != null && r.relation != relation) continue;
    if (r.resourceId != resourceId && r.resourceId != '*') continue;
    return true;
  }
  return false;
}

bool hasFgacRelation(
  List<FgacRelationClaim> relations,
  String relation,
  String resourceType,
  String resourceId,
) {
  return matchesFgacGrant(relations, resourceType, resourceId, relation: relation);
}

bool hasAnyFgacRelation(
  List<FgacRelationClaim> relations,
  Set<String> relationsToTry,
  String resourceType,
  String resourceId,
) {
  for (final rel in relationsToTry) {
    if (hasFgacRelation(relations, rel, resourceType, resourceId)) {
      return true;
    }
  }
  return false;
}

bool hasAllFgacRelations(
  List<FgacRelationClaim> relations,
  Set<String> relationsToTry,
  String resourceType,
  String resourceId,
) {
  for (final rel in relationsToTry) {
    if (!hasFgacRelation(relations, rel, resourceType, resourceId)) {
      return false;
    }
  }
  return true;
}

File: packages/ts/oidc-resource-server/src/claims.ts
import { z } from 'zod';

export const fgacRelationEntrySchema = z.object({
	resource_type: z.string(),
	resource_id: z.string(),
	relation: z.string(),
});

export type FgacRelationClaim = z.infer<typeof fgacRelationEntrySchema>;

export const fgacPermissionEntrySchema = z.object({
	resource_type: z.string(),
	resource_id: z.string(),
	permissions: z.array(z.string()),
});

export type FgacPermissionClaim = z.infer<typeof fgacPermissionEntrySchema>;

const realmAccessSchema = z.object({
	roles: z.array(z.string()),
});

const resourceAccessEntrySchema = z.object({
	roles: z.array(z.string()),
});

export function parseScopeClaim(scope: unknown): Set<string> {
	if (typeof scope !== 'string' || scope.trim().length === 0) {
		return new Set();
	}
	return new Set(scope.trim().split(/\s+/).filter(Boolean));
}

export function parseRealmRoles(claims: Record<string, unknown>): string[] {
	const raw = claims['realm_access'];
	const parsed = realmAccessSchema.safeParse(raw);
	if (!parsed.success) {
		return [];
	}
	return parsed.data.roles;
}

export function parseResourceAccess(claims: Record<string, unknown>): ReadonlyMap<string, readonly string[]> {
	const raw = claims['resource_access'];
	if (typeof raw !== 'object' || raw === null) {
		return new Map();
	}
	const out = new Map<string, readonly string[]>();
	for (const [k, v] of Object.entries(raw)) {
		const p = resourceAccessEntrySchema.safeParse(v);
		if (p.success) {
			out.set(k, p.data.roles);
		}
	}
	return out;
}

export function parseFgacRelations(claims: Record<string, unknown>): readonly FgacRelationClaim[] {
	const raw = claims['fgac_relations'];
	if (!Array.isArray(raw)) {
		return [];
	}
	const out: FgacRelationClaim[] = [];
	for (const item of raw) {
		const p = fgacRelationEntrySchema.safeParse(item);
		if (p.success) {
			out.push(p.data);
		}
	}
	return out;
}

export function parseFgacPermissions(claims: Record<string, unknown>): readonly FgacPermissionClaim[] {
	const raw = claims['fgac_permissions'];
	if (!Array.isArray(raw)) {
		return [];
	}
	const out: FgacPermissionClaim[] = [];
	for (const item of raw) {
		const p = fgacPermissionEntrySchema.safeParse(item);
		if (!p.success) {
			continue;
		}
		out.push({
			resource_type: p.data.resource_type,
			resource_id: p.data.resource_id,
			permissions: Array.from(new Set(p.data.permissions)),
		});
	}
	return out;
}

export function parseFgacTruncated(claims: Record<string, unknown>): boolean {
	return claims['fgac_truncated'] === true;
}

export function matchesFgacGrant(
	relations: readonly FgacRelationClaim[],
	resourceType: string,
	resourceId: string,
	relation?: string,
): boolean {
	for (const r of relations) {
		if (r.resource_type !== resourceType) {
			continue;
		}
		if (relation !== undefined && r.relation !== relation) {
			continue;
		}
		if (r.resource_id !== resourceId && r.resource_id !== '*') {
			continue;
		}
		return true;
	}
	return false;
}

export function relationsHeldOnResource(
	relations: readonly FgacRelationClaim[],
	resourceType: string,
	resourceId: string,
): Set<string> {
	const names = new Set<string>();
	for (const r of relations) {
		if (r.resource_type !== resourceType) {
			continue;
		}
		if (r.resource_id !== resourceId && r.resource_id !== '*') {
			continue;
		}
		names.add(r.relation);
	}
	return names;
}

export function hasFgacPermission(
	permissions: readonly FgacPermissionClaim[],
	resourceType: string,
	resourceId: string,
	permission: string,
): boolean {
	for (const entry of permissions) {
		if (entry.resource_type !== resourceType) {
			continue;
		}
		if (entry.resource_id !== resourceId && entry.resource_id !== '*') {
			continue;
		}
		if (entry.permissions.includes(permission)) {
			return true;
		}
	}
	return false;
}
